src/metabase/search/scoring.clj

```diff
@@ -127,30 +128,50 @@
                           [(str/lower-case (name model)) i])
                         search-config/searchable-models)))
 
-(defn- score-with-match
+(defn- text-score-with-match
   [query-string result]
   (when (seq query-string)
-    (score-with match-based-scorers
-                (tokenize (normalize query-string))
-                result)))
+    (text-score-with match-based-scorers
+                     (tokenize (normalize query-string))
+                     result)))
+
+(defn- pinned-score
+  [{:keys [collection_position]}]
+  (or (some-> collection_position -)
+      0))
 
 (defn- compare-score-and-result
-  "Compare [score result] pairs. Must return -1, 0, or 1. The score is assumed to be a vector, and will be compared in
-  order."
-  [[score-1 _result-1] [score-2 _result-2]]
+  "Compare maps of scores and results. Must return -1, 0, or 1. The score is assumed to be a vector, and will be
+  compared in order."
+  [{score-1 :score} {score-2 :score}]
   (compare score-1 score-2))
 
 (defn- serialize
-  "Massage the raw result from the DB into something more useful for the client"
-  [{:keys [collection_id collection_name name display_name] :as row} {:keys [score match column match-context] :as hit}]
-  (-> row
-      (assoc
-       :name           (if (or (= column :name) (nil? display_name)) name display_name)
-       :matched_column column
-       :matched_text   match
-       :context        (when-not (#{:name :display_name :collection_name} column) match-context) ;; TODO pull this out somewhere more responsible
-       :score          score
-       :collection     {:id collection_id :name collection_name})))
+  "Massage the raw result from the DB and match data into something more useful for the client"
+  [{:keys [result column match-context-thunk]}]
+  (let [{:keys [name display_name
+                collection_id collection_name]} result]
+    (-> result
+        (assoc
+         :name           (if (or (= column :name)
+                                 (nil? display_name))
+                           name
+                           display_name)
+         :context        (when-not (search-config/displayed-columns column)
+                           (match-context-thunk))
+         :collection     {:id   collection_id
+                          :name collection_name})
+        (dissoc
+         :collection_id
+         :collection_name
+         :display_name))))
+
```

Comment:
this style of assoc and then dissoc makes it confusing what is actually in this map. Is it possible to just return a map literal of all the keys? `{:name :context :collection ...}` Its not clear how much of `result` is still in this map.

Reply:
or maybe a schema map to type it (if possible. not a huge problem)

Reply:
A map literal isn't great since it's mostly passing through the keys. For example, the other open PR just added `dashboardcard_count` as another key and it's nice not to have to worry about it here.

I think about it as the `assoc` adding things that the FE wants: a nested object for the collection, better info about which name to use, and the match context (which lets matching words be bolded). The `dissoc` isn't strictly necessary, but I figured I may as well clean up a few things that were clearly not going to be used: the other format for collections, and the "you might think `display_name` was the name to display, but actually we just made it so regular `name` has you covered".

I supposed it'd be reasonable to make an allowlist of everything the client wants (currently just name, collection, and context), but @kdoh asked to return verbose results for now as the FE takes shape. For example, maybe he'll want to add an indicator that the result was pinned, or show how many dashboards it appears in, or whatever.

Reply:
ah ok. didn't know how complex the result was. fair enough. we can firm it up later if that's a desire but no need to now.